### PR TITLE
[SPARK-27324][DOC][CORE] Document configurations related to executor metrics and modify a configuration.

### PR DESCRIPTION
**Pull request by wypoon**
### What changes were proposed in this pull request?

Add a section to the Configuration page to document configurations for executor metrics.
At the same time, rename spark.eventLog.logStageExecutorProcessTreeMetrics.enabled to spark.executor.processTreeMetrics.enabled and make it independent of spark.eventLog.logStageExecutorMetrics.enabled.

### Why are the changes needed?

Executor metrics are new in Spark 3.0. They lack documentation.
Memory metrics as a whole are always collected, but the ones obtained from the process tree have to be optionally enabled. Making this depend on a single configuration makes for more intuitive behavior. Given this, the configuration property is renamed to better reflect its meaning.

### Does this PR introduce any user-facing change?

Yes, only in that the configurations are all new to 3.0.

### How was this patch tested?

Not necessary.

### CODE REVIEW

**Comment by wypoon** [comment]
@rezasafi I'd like to propose renaming `spark.eventLog.logStageExecutorProcessTreeMetrics.enabled` to say, `spark.executor.processTreeMetrics.enabled`. Unlike `spark.eventLog.logStageExecutorMetrics.enabled`, which determines whether to **log** executor metrics (which are always collected), this property determines whether to collect the process tree metrics in the first place. As such, I think the current name is slightly misleading and could be improved. Since all these configurations are new in 3.0, I'd like to do the renaming now, before 3.0 is released. What do you think?
@squito ?

**Comment by rezasafi** [comment]
@wypoon processtree metrics will also be written to eventlogs like other executor metrics and they will be gathered only if spark.eventLog.logStageExecutorMetrics.enabled is set to true. I think the current name which is "spark.eventLog.logStageExecutorProcessTreeMetrics.enabled" is self explanatory about this behavior. So If you want to change the name it makes more sense to change "spark.eventLog.logStageExecutorMetrics.enabled" as well. OtherwiseI think the current name is better to avoid user mis understanding.

**Comment by wypoon** [comment]
> @wypoon processtree metrics will also be written to eventlogs like other executor metrics and they will be gathered only if spark.eventLog.logStageExecutorMetrics.enabled is set to true. I think the current name which is "spark.eventLog.logStageExecutorProcessTreeMetrics.enabled" is self explanatory about this behavior. So If you want to change the name it makes more sense to change "spark.eventLog.logStageExecutorMetrics.enabled" as well. OtherwiseI think the current name is better to avoid user mis understanding.

It is not self-evident that spark.eventLog.logStageExecutorProcessTreeMetrics.enabled being true is only effective if spark.eventLog.logStageExecutorMetrics.enabled is true as well. This is one reason why we need documentation and I document this fact in this PR.
As I already pointed out, spark.eventLog.logStageExecutorMetrics.enabled is about whether to log metrics - the metrics (in general) are always collected - so that name is appropriate. spark.eventLog.logStageExecutorProcessTreeMetrics.enabled is about whether to collect a particular class of metrics in the first place. If they are not collected, they are logged as 0 (but still logged), if spark.eventLog.logStageExecutorMetrics.enabled is true.
This is why I propose the rename.

**Comment by rezasafi** [comment]
I think when the name of the two options have the same prefix naturally it is more clear that the two metrics have some sort of a relation. I really not sure how much benefit a rename will have. 
Also your jira is about documentation and it seems weird to do a rename under that, however it isn't a weirdest thing that I have seen in this project!

**Comment by srowen** [comment]
No strong opinion on this here. 
If the metric being remove was never present in 2.x then it's OK, otherwise I think we'd have to deprecate, etc.

**Review by srowen — COMMENTED** [review]
yes the change is straightforward, but what about `spark.eventLog.logStageExecutorMetrics.enabled`? Does it need to change too or is that pretty different?

**Comment by wypoon** [comment]
> yes the change is straightforward, but what about `spark.eventLog.logStageExecutorMetrics.enabled`? Does it need to change too or is that pretty different?

That property does relate directly to the event log; it determines if aggregated metrics are written to the log. I did not feel the need to change that.

**Comment by dongjoon-hyun** [comment]
@wypoon Could you update the PR title since this PR touches the code?

**Comment by wypoon** [comment]
> @wypoon Could you update the PR title since this PR touches the code?

Done.

**Comment by squito** [comment]
from a quick look, I think this makes sense (including the rename), but I need sometime to look back through the code and see where exactly these properties are touched to make sure.  It does seem like the property has more to do with whether executors collect the metrics.  Writing to the event log does naturally follow, btu you also get the metrics in any listener.

Also cc @LucaCanali who might have some thoughts here.

File: docs/configuration.md [hunk 1]
```diff
@@ -1624,6 +1624,42 @@ Apart from these, the following properties are also available, and may be useful
...
+    exists and <code>spark.eventLog.logStageExecutorMetrics.enabled</code> is true as well.
+  </td>
+<tr>
+  <td><code>spark.executor.metrics.pollingInterval</code></td>
+  <td>0</td>
+  <td>
+    How often to collect executor metrics.
+    <br />
+    If 0, the polling is done on executor heartbeats (thus at the heartbeat interval).
+    If positive, the polling is done at this interval.
```
Comment (LucaCanali) [inline docs/configuration.md]: I propose to add that the polling interval is expressed in milliseconds

File: docs/configuration.md [hunk 2]
```diff
@@ -1624,6 +1624,42 @@ Apart from these, the following properties are also available, and may be useful
...
+    <em>Note:</em> The process tree metrics are collected only if the /proc filesystem
+    exists and <code>spark.eventLog.logStageExecutorMetrics.enabled</code> is true as well.
+  </td>
+<tr>
+  <td><code>spark.executor.metrics.pollingInterval</code></td>
+  <td>0</td>
+  <td>
+    How often to collect executor metrics.
+    <br />
+    If 0, the polling is done on executor heartbeats (thus at the heartbeat interval).
```
Comment (LucaCanali) [inline docs/configuration.md]: We could explicitly reference the parameter spark.executor.heartbeatInterval

**Comment by LucaCanali** [comment]
I think it is a good idea to add info on executor metrics configuration to `docs/configuration.md` as proposed in this PR.

The rationale discussed above for renaming `spark.eventLog.logStageExecutorProcessTreeMetrics.enabled` to `spark.executor.processTreeMetrics.enabled` seems valid to me. 

Additional comments:  
- This discussion actually exposes the strange relationship that the configuration `spark.eventLog.logStageExecutorProcessTreeMetrics.enabled` (here renamed to `spark.executor.processTreeMetrics.enabled`) has with `spark.eventLog.logStageExecutorMetrics.enabled`. Maybe there should not be the dependency that process tree metrics are collecteded if both `spark.eventLog.logStageExecutorMetrics.enabled` and `spark.executor.processTreeMetrics.enabled` are set to true, but we could change that and just use the parameter `spark.executor.processTreeMetrics.enabled` for this scope?

- Related to this discussion, I have also noticed that the description of `spark.eventLog.logStageExecutorMetrics.enabled` in `docs/monitoring.md` needs to be clarified, as it currently states "Metrics `peakExecutorMetrics.*` are only enabled if `spark.eventLog.logStageExecutorMetrics.enabled` is true".   
A change to that text could be added to this PR, following the parameter description in `configurations.md`?

**Comment by wypoon** [comment]
@LucaCanali thanks for your comments.
At this point, I do not wish to change the behavior. Renaming a configuration property is one thing; it is a simple change. Changing the effect of a property is another thing, and requires testing. You can understand if I don't wish to take that on.
On your last point, I don't know if the statement in question is inaccurate (as I haven't examined that type of metrics). What I can clarify is that the executor always polls memory metrics (either in the heartbeat thread or in a separate thread) and sends them in its heartbeat. In the `EventLoggingListener`, there is logic to aggregate these memory metrics and log per-stage per-executor peaks on stage completion, but only if `spark.eventLog.logStageExecutorMetrics.enabled` is true (otherwise executor metrics update events are ignored).
Just to confirm, whenever the `ExecutorMetricsPoller` is asked to `poll`, then you update your `ExecutorMetricsSource` with the latest snapshot of the memory metrics, correct? In this case, this happens irrespective of  `spark.eventLog.logStageExecutorMetrics.enabled`. If this is correct, that the statement you refer to in `docs/monitoring.md` is inaccurate and should be corrected.

**Comment by wypoon** [comment]
@LucaCanali thinking about it a bit more, if my understanding is correct, and memory metrics (except for process tree metrics) are always collected for the Dropwizard metrics, then I understand your wanting the behavior change you suggest (not have `spark.executor.processTreeMetrics.enabled` depend on `spark.eventLog.logStageExecutorMetrics.enabled`).

**Comment by squito** [comment]
ok, yes I also see @rezasafi 's point now.  The procfs metrics are only gathered if event log stage executor metrics are turned on:

https://github.com/apache/spark/blob/f59685acaa3e9c227f14fe4d8f9e94a1ac664b05/core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala#L61-L65

Which makes the current name more reasonable.  But that is strange behavior, particularly with https://issues.apache.org/jira/browse/SPARK-27189 -- I should have thought to change that at the same time.

@wypoon I do think that there should also be behavior change as part of this, to remove that check in procfsmetricsgetter.

**Comment by LucaCanali** [comment]
As far as I understand, also tested by running a few tests, the current situation is:  
As you wrote, `ExecutorMetricsSource` reports executor memory metrics values from the latest snapshot of the metrics as sent by the `ExecutorMetricsPoller`. These are not peak values of the metrics but rather represent the time evolution (for monitoring pourposes, handled via the dropwizard library, needs to be configured with the relevant parameters).  
Peak values of the executor memory metrics are reported by the Web UI under /api/v1/applications/.../executors for each executor. This information is made available irrespective of the value of `spark.eventLog.logStageExecutorMetrics.enabled`  
`SparkListenerStageExecutorMetrics` are logged to the eventlog with per-stage computed peak values of the executor metrics only if `spark.executor.processTreeMetrics.enabled` is set to true, as you wrote.

**Comment by wypoon** [comment]
Thanks @LucaCanali and @squito. I will change the behavior in `ProcfsMetricsGetter` so collection of proc fs metrics do not depend on `spark.eventLog.logStageExecutorMetrics.enabled`, only on (the renamed) `spark.executor.processTreeMetrics.enabled`.

File: docs/configuration.md (hunk 1 above)
Comment (wypoon) [reply docs/configuration.md]: Added unit.

File: docs/configuration.md (hunk 2 above)
Comment (wypoon) [reply docs/configuration.md]: Added reference.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 3]
```diff
@@ -233,6 +228,11 @@ package object config {
   private[spark] val EXECUTOR_HEARTBEAT_MAX_FAILURES =
     ConfigBuilder("spark.executor.heartbeat.maxFailures").internal().intConf.createWithDefault(60)
 
+  private[spark] val EXECUTOR_PROCESS_TREE_METRICS_ENABLED =
+    ConfigBuilder("spark.executor.processTreeMetrics.enabled")
+      .booleanConf
+      .createWithDefault(false)
```
Comment (rezasafi) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: shouldn't we add .doc for this metric and also spark.eventLog.logStageExecutorMetrics.enabled in addition to what is added to configuration.md?

**Comment by rezasafi** [comment]
The goal of SPARK-24958, which introduces procfs metrics, was to compute peak procefs metrics along with other memory metrics and log them in event logs. So because of that there was a need that spark.eventLog.logStageExecutorMetrics.enabled being true when you want procfs metrics. 
I wasn't aware of SPARK-27189 when I objected the rename. After SPARK-27189  this has been changed. The metric system doesn't deal with peak values, but it exposed the processtree metrics. So the rename make sense now to me. 
Still I think we need to document somewhere that if you need peak metrics  for processtree metrics, you need to set both spark.eventLog.logStageExecutorMetrics.enabled and whatever name you decided for enabling for processtree metrics.

**Comment by wypoon** [comment]
> Still I think we need to document somewhere that if you need peak metrics for processtree metrics, you need to set both spark.eventLog.logStageExecutorMetrics.enabled and whatever name you decided for enabling for processtree metrics.

I think this is clear from the documentation I added.

**Comment by squito** [comment]
lgtm assuming tests pass & add the `.doc` as suggested by reza above.

**Comment by rezasafi** [comment]
> > Still I think we need to document somewhere that if you need peak metrics for processtree metrics, you need to set both spark.eventLog.logStageExecutorMetrics.enabled and whatever name you decided for enabling for processtree metrics.
> 
> I think this is clear from the documentation I added.

yeah, I can see that. lgtm

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 3 above)
Comment (wypoon) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Added.

**Comment by wypoon** [comment]
Argh, I misread the metrics (in `docs/monitoring.md`) @LucaCanali referred to in his first comment. I though he was referring to the Dropwizard metrics he worked on, but actually that section was added by Lantao Jin as documentation for what Edwina Lu added. So I mistakenly deleted a sentence from that section, which I have now restored.
Sorry for the confusion.

**Comment by LucaCanali** [comment]
Thanks @wypoon for the changes, LGTM.
In case you are interested, I would like to propose a small change to monitoring.md that I think may be a good fit here as an addition/complement to the PR (or I can do a separate PR for that, if preferred). At line 643, Delete the current text and replace it with an updated one with more details:

Delete:
"""
Metrics `peakExecutorMetrics.*` are only enabled if `spark.eventLog.logStageExecutorMetrics.enabled` is true. 
"""

Replace with:
"""
Executor metric values and their measured peak values per executor are exposed via the REST API at the end point `/applications/[app-id]/executors`.
In addition, aggregated per-stage peak values of the executor metrics are written to the event log if `spark.eventLog.logStageExecutorMetrics.enabled` is true.
Executor metrics are also exposed via the Spark metrics system based on the Dropwizard metrics library.
"""

**Comment by wypoon** [comment]
@LucaCanali thank you for the clarifying text. I shall do it.

**Comment by squito** [comment]
merged to master, thanks everyone!

**Comment by wypoon** [comment]
Hmm, between the last passing test build (117600) and the next one that failed (117680), there was no code change, only a change in docs/monitoring.md. Thus the test failure cannot be caused by my final commit. Perhaps the test is flaky?

**Pull request closed without merging**